Desktop, Android: Fix historic issue whereby the first revision created for a note does not contain the original contents

mrjo118 opened the pull request:
**Problem:**
When a note which does not have any revisions is modified, when the maintenance kicks in, the first revision which is created does not contain the original content of the note, but instead contains a snapshot of the content of the note at some point within the first 10 minutes of editing, depending on when the maintenance kicks in.

The issue occurs regardless of whether the updated_time is before or after the old note cut off date defined by the 'revisionService.oldNoteInterval' setting which is hardcoded to 7 days. This is particularly disturbing in the case of very old notes. Take for example, I have a note which I have not modified in 2 years. When reading it, I highlight a paragraph of text and accidentally replace it with some gibberish characters without noticing. The next day I come back to the note and realise what I have done. I refer to the note history and I find there is only 1 revision, and that revision contains the mistake I made, so I have no way to go back to the old version. I would have expected better from a note history feature.

Reproduction steps on Joplin desktop:
1. Take a note which has no version history
2. Add some text to the note, and wait a few seconds
3. Restart Joplin
4. Wait a few seconds, or check the log until the revision maintenance has completed
5. Observe that 1 revision has been created, but it does not contain the original note contents, instead it contains the updated content

This PR addresses the issues by doing the following:
1. On Note.save, it currently checks whether the current updated_time on a note is before the cut off date, and stores an isOld boolean in a cache on the RevisionService (isOldNotesCache_). If the value evaluates to true, it will set the value of beforeNoteJson to the old note entity. There are 2 problems with this:
- Currently the logic does not await the result of this.revisionService().isOldNote(o.id)
- If it does evaluate to true, then if save is called multiple times consecutively (i.e. if the user is continually typing, or makes multiple changes before collectRevisions kicks in), then the item_changes record will get replaced with each save, which results in the loss of the original content. In order to remedy this, I have changed the code so that instead of caching a boolean for isOld, it instead caches the oldNote object, and will set the cached value to beforeNoteJson to avoid losing the original content
2. Even when remedying the first point, notes which have been changed within the last 7 days but have no note history will exhibit this issue. In particular this could be a problem now that we support deleting all revisions for an individual note. To address this, I have removed the isOld value from the cache in point 1 and now always set beforeNoteJson to the cache, rather than just when isOld is true. This does not impact performance, because we already select the old note on every save anyway, regardless of whether isOld is true or not. The code still compares against the oldNoteInterval in the collectRevisions function (as per existing), but in addition to this, the logic to create a revision using the oldNote now applies as well when the note has no history. If the note has only just been created though, we don't want to store empty note content in the first revision, so to avoid this we just need to check that the difference between the updated_time on the cached old note and the current updated_time is at least as long as the intervalBetweenRevisions value
3. The logic to create a revision using the old note is currently flawed, in that the updated_time on the old note is passed onto the item_change_date on the created revision. This means that if for example, the user had their note history expiry set to 90 days and made a single change to a note which was not modified for 2 years, the revisions would immediately get removed by the revision cleaner after creating it. Therefore we should instead use the current updated_time for the note to set as item_change_date, to avoid this. Note that this will not result in creating 2 revisions with exactly the same timestamp, because the intervalBetweenRevisions will prevent 2 revisions from being created within the same iteration of the loop in collectRevisions

**Testing:**
I have tested the following things, waiting a few seconds and restarting Joplin after each of the steps, to give to specified expected outcome:
1. Create a new note and add 3 characters to the content 1 at a time, entering the characters 2 seconds apart - EXPECTATION: 1 revision is created and it contains the current note content
2. Take an existing note created over 10 minutes ago and delete all the revisions. Add an additional character to the content - EXPECTATION: 1 revision is created and it contains the old note content
3. Take an existing note created over 10 minutes ago and delete all the revisions. Add 3 additional characters to the content 1 at a time, entering the characters 2 seconds apart - EXPECTATION: 1 revision is created and it contains the old note content
4. Take an existing note created over 7 days ago and delete all the revisions. Add 3 additional characters to the content 1 at a time, entering the characters 2 seconds apart - EXPECTATION: 1 revision is created and it contains the old note content
5. Take an existing note with at least 1 existing revision which is at least 10 minutes old. Add 3 additional characters to the content 1 at a time, entering the characters 2 seconds apart - EXPECTATION: 1 additional revision is created and it contains the current note content
6. After performing scenarios 3 and 5, wait 10 minutes, then add 3 additional characters to the content 1 at a time. Then wait up to 10 minutes for collectRevisions to kick in - EXPECTATION: 1 additional revision is created for each note and in both cases they contain the current note content

laurent22 commented:
Thanks for looking into this, but we also need test units for this. If it wasn't working before, it should be possible to create a failing tests to show what the issue was, and to ensure there is no regressions in the future

mrjo118 commented:
The pipelines have been flaky lately, but I did see some genuine test failures in the RevisionService tests on the first push I made which triggered the pipelines.

I'll look into updating those and improving the coverage.

mrjo118 commented:
> Thanks for looking into this, but we also need test units for this. If it wasn't working before, it should be possible to create a failing tests to show what the issue was, and to ensure there is no regressions in the future

@laurent22 it was a big job (and the PR is now significantly larger with changes to tests), but I've now fixed all the existing tests including comments / explanations and added additional test coverage to cover my testing scenarios. I increased the coverage of my manual testing as well since the initial state of the PR, which highlighted an issue that I have since remedied.

I've updated the PR description with updated analysis, test scenarios and a summary of my changes to the existing test coverage.

mrjo118 commented:
On further reflection, I realise that my code is caching the old note content in memory until the next revision collection, which is a bad idea. As the user may change many large notes before the next revision collection, which will result in high memory usage.

I'll need to rework this a bit, but I should be able to resolve this I think by caching a boolean for isOld similar to before, but where it is true, select the old note contents from the latest item_change, to set as the beforeItemChange value defined in Note.save

mrjo118 commented:
> I'll need to rework this a bit, but I should be able to resolve this I think by caching a boolean for isOld similar to before, but where it is true, select the old note contents from the latest item_change, to set as the beforeItemChange value defined in Note.save

I've now implemented this change. So the PR is ready for review again

personalizedrefrigerator commented on packages/app-mobile/components/screens/NoteRevisionViewer.test.tsx:
```diff
@@ -45,7 +45,7 @@ const createNoteWithTestRevisions = async (count: number) => {
 	// Verify that the revisions were created successfully
-	expect(await Revision.allByType(ModelType.Note, noteId)).toHaveLength(count);
+	expect(await Revision.allByType(ModelType.Note, noteId)).toHaveLength(count + 1);
 	return note;
```
**Optional/testing logic**: If it doesn't break the other tests, consider instead creating one fewer revision in the `for` loop just above this statement. Currently, the function is titled `createNoteWithTestRevisions` and accepts a `count: number`. It may make sense to have this function continue to result in just `count` revisions.

personalizedrefrigerator commented on packages/lib/services/RevisionService.ts:
```diff
@@ -46,12 +47,13 @@ export default class RevisionService extends BaseService {
-		const isOld = await Note.noteIsOlderThan(noteId, this.oldNoteCutOffDate_());
-		this.isOldNotesCache_[noteId] = isOld;
-		return isOld;
+		// No particular value needs to be stored, but it is more performant to look up a key on an object than searching for a value in an array
+		this.changedSinceCollectionCache_[noteId] = null;
+
```
> it is more performant to look up a key on an object than searching for a value in an array

**Note**: An alternative here might be to make `changedSinceCollectionCache_` a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set). However, this may require more refactoring.

mrjo118 commented:
@personalizedrefrigerator I've addressed both comments, and the relevant tests still pass locally

mrjo118 commented:
Closing this in favour of https://github.com/laurent22/joplin/pull/12674
